plugins/dell-dock2/fu-dell-dock2-rtshub.c

```diff
@@ -0,0 +1,596 @@
+			    (const guint8 *)&cmd_buffer,
+			    sizeof(cmd_buffer),
+			    0x0, /* src */
+			    sizeof(cmd_buffer),
+			    error))
+		return FALSE;
+	if (!fu_hid_device_set_report(FU_HID_DEVICE(self),
+				      0x0,
+				      buf,
+				      sizeof(buf),
+				      DELL_DOCK2_RTSHUB_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_NONE,
+				      error)) {
+		g_prefix_error(error, "failed to soft reset: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock2_rtshub_write_flash(FuDellDock2RtsHub *self,
+				 guint32 addr,
+				 const guint8 *data,
+				 guint16 data_sz,
+				 GError **error)
+{
+	FuRtshubHIDCmdBuffer cmd_buffer = {
+	    .cmd = RTSHUB_CMD_WRITE_DATA,
+	    .ext = RTSHUB_EXT_WRITEFLASH,
+	    .dwregaddr = GUINT32_TO_LE(addr),
+	    .bufferlen = GUINT16_TO_LE(data_sz),
+	};
+	guint8 buf[DELL_DOCK2_RTSHUB_BUFFER_SIZE] = {0};
+
+	g_return_val_if_fail(data_sz <= 128, FALSE);
+	g_return_val_if_fail(data != NULL, FALSE);
+	g_return_val_if_fail(data_sz != 0, FALSE);
+
+	/* vendor command */
+	if (!fu_memcpy_safe(buf,
+			    sizeof(buf),
+			    0x0, /* dst */
+			    (const guint8 *)&cmd_buffer,
+			    sizeof(cmd_buffer),
+			    0x0, /* src */
+			    sizeof(cmd_buffer),
+			    error))
+		return FALSE;
+
+	/* data payload */
+	if (!fu_memcpy_safe(buf,
+			    sizeof(buf),
+			    DELL_DOCK2_RTSHUB_WRITE_FLASH_OFFSET_DATA, /* dst */
+			    data,
+			    data_sz,
+			    0x0, /* src */
+			    data_sz,
+			    error))
+		return FALSE;
+
+	if (!fu_hid_device_set_report(FU_HID_DEVICE(self),
+				      0x0,
+				      buf,
+				      sizeof(buf),
+				      DELL_DOCK2_RTSHUB_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_NONE,
+				      error)) {
+		g_prefix_error(error, "failed to write flash @%08x: ", (guint)addr);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gchar *
+fu_dell_dock2_rtshub_substr_before_colon(const gchar *str)
+{
+	gchar *colon = g_strrstr(str, ":");
+	return (colon != NULL) ? g_strndup(str, colon - str) : g_strdup(str);
+}
+
+static gboolean
+fu_dell_dock2_rtshub_is_sibling(FuDevice *device, FuDevice *sibling)
+{
+	const gchar *dev_phyid = fu_device_get_physical_id(device);
+	const gchar *sib_phyid = fu_device_get_physical_id(sibling);
+	const gchar *dev_phyid_prefix = fu_dell_dock2_rtshub_substr_before_colon(dev_phyid);
+	const gchar *sib_phyid_prefix = fu_dell_dock2_rtshub_substr_before_colon(sib_phyid);
+
+	/* compare with physical id */
+	return g_strcmp0(dev_phyid_prefix, sib_phyid_prefix) == 0;
+}
+
+static gboolean
+fu_dell_dock2_rtshub_care_family_back(FuDevice *device, GError **error)
+{
+	FuDellDock2RtsHub *self = FU_DELL_DOCK2_RTSHUB(device);
+	FuDevice *parent = fu_device_get_parent(device);
+
+	/* wait myself back */
+	fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	g_debug("device (%s) wants to come back after reset", fu_device_get_name(device));
+
+	/* wait parent back */
+	if (self->ensure_parent_after_reset) {
+		fu_device_add_flag(parent, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
```

Comment:
We know that the parent device will return, and it did successfully.  Nevertheless, the old device doesn't hand off the parentalship to the new device. The negative impact is that all the children will be referring to a dead `disonnected` device of parent.

The use case in this code, is that we need `parent` to do something in the `write_firmware` lifecycle. That'll end up referring to a `disconnected` parent.

RTSHUB Gen 1 -> use parent do something -> update done -> wait for parent back -> parent is back.
RTSHUB Gen 2 -> use parent do something -> parent already disconnected.

What is the suggestion to resolve this issue?  We want hard code the relationship in device code here e.g.: rtshub device to rebind the relationship at all best. Or we should fix it in the fu_device_list area?

Reply:
Added screenshots and runtime log. The `0x55555580cc90` at write-firmware runtime is actually an old parent device.

```
  25009 6 [0x5555557e9090] 
  25010 new: b7411b4ba4a3edb55b34597dd9f452eacc076f5d [0x555555afdbb0]
  25011 old: b7411b4ba4a3edb55b34597dd9f452eacc076f5d [0x55555580cc90]
```
![Screenshot from 2024-07-30 18-21-59](https://github.com/user-attachments/assets/87e6ae97-a4ee-4bf8-86d0-119134590a69)
![Screenshot from 2024-07-30 18-20-47](https://github.com/user-attachments/assets/151896dd-abce-426d-a682-892b389b37cf)
[x.log](https://github.com/user-attachments/files/16426197/x.log)

Reply:
Proposed a fix in https://github.com/fwupd/fwupd/pull/7556.